bootstore/src/schemes/v0/request_manager.rs

```diff
@@ -0,0 +1,401 @@
+    ///
+    /// Only peers in `InitialMember` or `Learned` state can load rack secrets
+    LoadRackSecret { rack_uuid: Uuid, acks: ShareAcks },
+
+    /// A request received from a peer to learn a new share
+    ///
+    /// This request results in share gathering so "extra" shares can be
+    /// decrypted via a key derived from the rack secret and one of them
+    /// distributed to the learner.
+    ///
+    /// Only peers in `InitialMember` state can respond successfully
+    LearnReceived { rack_uuid: Uuid, from: Baseboard, acks: ShareAcks },
+
+    /// A request sent from a peer in `Learning` state to another peer
+    /// to learn a key share.
+    LearnSent { to: Baseboard },
+}
+
+/// A mechanism to manage all in flight requests
+///
+/// We expect very few requests at a time - on the order of one or two requests.
+pub struct RequestManager {
+    id: Baseboard,
+    config: Config,
+    requests: BTreeMap<Uuid, TrackableRequest>,
+    expiry_to_id: BTreeMap<Instant, Uuid>,
+
+    /// Messages that need sending to other peers.
+    ///
+    /// These should be drained on each API call.
+    envelopes: Vec<Envelope>,
+}
+
+impl RequestManager {
+    /// Create a new RequestManager
+    pub fn new(id: Baseboard, config: Config) -> RequestManager {
+        RequestManager {
+            id,
+            config,
+            requests: BTreeMap::new(),
+            expiry_to_id: BTreeMap::new(),
+            envelopes: vec![],
+        }
+    }
+
+    /// Return an iterator of drained envelopes
+    pub fn drain_elements(&mut self) -> impl Iterator<Item = Envelope> + '_ {
+        self.envelopes.drain(..)
+    }
+
+    /// Track a new `Fsm::init_rack` api request and broadcast a
+    /// `RequestType::Init` to connected peers.
+    pub fn new_init_rack_req(
+        &mut self,
+        now: Instant,
+        rack_uuid: Uuid,
+        packages: BTreeMap<Baseboard, SharePkg>,
+        connected_peers: &BTreeSet<Baseboard>,
+    ) -> Uuid {
+        let expiry = now + self.config.rack_init_timeout;
+        let mut acks = InitAcks::default();
+        acks.expected = packages
+            .keys()
+            .cloned()
+            .filter_map(|id| if id != self.id { Some(id) } else { None })
+            .collect();
+        let req = TrackableRequest::InitRack {
+            rack_uuid,
+            packages: packages.clone(),
+            acks,
+        };
+        let request_id = self.new_request(expiry, req);
+
+        // Send a `Request::Init` to each connected peer in the initial group
+        let iter = packages.into_iter().filter_map(|(to, pkg)| {
+            if connected_peers.contains(&to) {
+                Some(Envelope {
+                    to,
+                    msg: Request {
+                        id: request_id,
+                        type_: RequestType::Init(pkg),
+                    }
+                    .into(),
+                })
+            } else {
+                None
+            }
+        });
```

Comment:
I'm not sold on the `filter_map` here either
```suggestion
        let iter = packages
            .into_iter()
            .filter(|(to, _pkg)| connected_peers.contains(to))
            .map(|(to, pkg)| Envelope {
                to,
                msg: Request { id: request_id, type_: RequestType::Init(pkg) }
                    .into(),
            });
```

Reply:
Agreed. Your version is clearer. Thanks!